# origin: chrisw (BitKeeper)
# cset: 1.1485 (2.4) key=4244a417jb-ZTWgax2mcR2so3l3KaQ
# URL: http://linux.bkbits.net:8080/linux-2.4/cset@4244a417jb-ZTWgax2mcR2so3l3KaQ
# inclusion: upstream
# descrition: [PATCH] isofs: Some more defensive checks to keep corrupt isofs images from corrupting memory/oopsing.
# revision date: Tue, 29 Mar 2005 17:40:41 +0900
#
# S rset: ChangeSet|1.1484..1.1485
# I rset: fs/isofs/rock.c|1.8..1.9
#
# Key:
# S: Skipped  ChangeSet file only
# O: Original Followed by Updated
# U: Updated  Included with updated range of versions
# I: Included Included verbatim
# E: Excluded Excluded on request from user
# D: Deleted  Manually deleted by subsequent user edit
# R: Revised  Manually revised by subsequent user edit
#
#
# This is a BitKeeper generated diff -Nru style patch.
#
# fs/isofs/rock.c
#   2005/03/24 05:41:31-03:00 chrisw@osdl.org +4 -0
#   isofs: Some more defensive checks to keep corrupt isofs images from corrupting memory/oopsing.
# 
# ChangeSet
#   2005/03/25 20:51:51-03:00 chrisw@osdl.org 
#   [PATCH] isofs: Some more defensive checks to keep corrupt isofs images from corrupting memory/oopsing.
#   
#   Michal Zalewski <lcamtuf@dione.ids.pl> discovers range checking flaws in
#   iso9660 filesystem.
#   
#   http://marc.theaimsgroup.com/?l=bugtraq&m=111110067304783&w=2
#   
#   CAN-2005-0815 is assigned to this issue.
#   
#   Some more defensive checks to keep corrupt isofs images from corrupting
#   memory or causing Oops.
#   
#   Signed-off-by: Chris Wright <chrisw@osdl.org>
#   
#   ===== fs/isofs/rock.c 1.23 vs edited =====
# 
#
===== fs/isofs/rock.c 1.8 vs 1.9 =====
--- 1.8/fs/isofs/rock.c	2005-03-26 12:38:11 +09:00
+++ 1.9/fs/isofs/rock.c	2005-03-24 17:41:31 +09:00
@@ -73,6 +73,10 @@
     offset1 = 0; \
     pbh = sb_bread(DEV->i_sb, block); \
     if(pbh){       \
+      if (offset > pbh->b_size || offset + cont_size > pbh->b_size){	\
+	brelse(pbh); \
+	goto out; \
+      } \
       memcpy(buffer + offset1, pbh->b_data + offset, cont_size - offset1); \
       brelse(pbh); \
       chr = (unsigned char *) buffer; \
